### PR TITLE
Deprecate the omega tactic

### PR DESCRIPTION
**Pull request by vbgl**
This deprecates the `omega` tactic. Tactics from the micromega plug-in can be used instead.

This adds a new file: what header/copyright/disclaimer/ascii-art logo should I put in it?

### CODE REVIEW

**Comment by Zimmi48** [comment]
> This adds a new file: what header/copyright/disclaimer/ascii-art logo should I put in it?

Just copy the content of `dev/header.ml`

File: doc/changelog/04-tactics/11976-deprecate-omega.rst
Comment (Zimmi48) [inline doc/changelog/04-tactics/11976-deprecate-omega.rst]:
```suggestion
  The :tacn:`omega` tactic is deprecated;
```

File: doc/changelog/04-tactics/11976-deprecate-omega.rst
Comment (Zimmi48) [inline doc/changelog/04-tactics/11976-deprecate-omega.rst]: if the proposed replacement is always the tactic lia then:
```suggestion
  use :tacn:`lia` from the :ref:`micromega` plugin instead
```

File: doc/sphinx/addendum/omega.rst
Comment (Zimmi48) [inline doc/sphinx/addendum/omega.rst]:
```suggestion
   .. deprecated:: 8.12

      Use :tacn:`lia` instead.
```

You should also adjust the warning above, which currently says that `omega` is "about to be deprecated".

File: doc/sphinx/addendum/omega.rst
Comment (Zimmi48) [inline doc/sphinx/addendum/omega.rst]: What about turning the warning off in the examples?

File: doc/sphinx/proof-engine/ltac.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/ltac.rst]: What about replacing `omega` with `lia` in this example. Will this break the example because `lia` performs much better. cc @JasonGross who is probably the author of this example.

**Review by Zimmi48 — APPROVED** [review]

File: doc/changelog/04-tactics/11976-deprecate-omega.rst
Comment (vbgl) [reply doc/changelog/04-tactics/11976-deprecate-omega.rst]: The rendering would be weird: the link text is very long.

File: doc/sphinx/addendum/omega.rst
Comment (vbgl) [reply doc/sphinx/addendum/omega.rst]: Done. Thanks.

File: doc/sphinx/addendum/omega.rst
Comment (vbgl) [reply doc/sphinx/addendum/omega.rst]: I think it’s fine to show the warning in the manual.

File: doc/changelog/04-tactics/11976-deprecate-omega.rst
Comment (Zimmi48) [reply doc/changelog/04-tactics/11976-deprecate-omega.rst]: Ah OK, then this could be:
```suggestion
  use :tacn:`lia` from the :ref:`micromega <micromega>` plugin instead
```

Comment (vbgl) [reply doc/changelog/04-tactics/11976-deprecate-omega.rst]: Excellent!

File: doc/sphinx/proof-engine/ltac.rst
Comment (vbgl) [reply doc/sphinx/proof-engine/ltac.rst:1732]: IMHO, we can leave this example as-is for now. This does not prevent Jason or anyone to rework it in a separate PR.

**Comment by vbgl** [comment]
This is ready…

File: doc/sphinx/proof-engine/ltac.rst
Comment (Zimmi48) [reply doc/sphinx/proof-engine/ltac.rst:1735]: OK to leave `omega` in there, but you should locally disable the warning. Currently, the output contains:

```
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    Toplevel input, characters 0-4:
    > tac.
    > ^^^^
    Warning: omega is deprecated since 8.12; use “lia” instead.
    [omega-is-deprecated,deprecated]
    No more subgoals.
```

It seriously undermine the readability of the example.

**Review by Zimmi48 — COMMENTED** [review]
It seems to me that importing `ZArith_hints` should be advertised as a replacement for importing `Omega` to enhance `auto with zarith`. Another option would be to have this `ZArith_hints` module be defined within `Lia` so that `Require Import Lia` enhances `auto with zarith` (although this could be a breaking change and I'm not sure we'd want this).

**Comment by Zimmi48** [comment]
What about the objective of closing the last omega / lia comparison issues first?

There is this example:

```coq
Goal let t' := nat in forall (n : t'), n = n.
Proof.
  intros.
  Fail lia.
  omega.
Qed.
```

from #11089 (we should decide whether to support this or close as "won't fix").

And there is #9921 which, according to its own author, is "not a blocker anymore". We can also decide to leave that for further performance improvements.

File: doc/sphinx/proof-engine/ltac.rst
Comment (vbgl) [reply doc/sphinx/proof-engine/ltac.rst]: Done.

**Comment by vbgl** [comment]
> It seems to me that importing ZArith_hints should be advertised as a replacement for importing Omega to enhance auto with zarith.

You can import `ZArith`, as before, and get the same behavior.

I think that ZArith_hint is an internal module that does not deserve to be explicitly documented.

**Comment by Zimmi48** [comment]
> You can import `ZArith`, as before, and get the same behavior.

OK, I hadn't realized that `ZArith` still exports `Omega`, which export `ZArith_hint`.

File: doc/sphinx/proof-engine/vernacular-commands.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/vernacular-commands.rst:255]: This line shouldn't be necessary since `ZArith` exports `Omega` which exports `OmegaLemmas`. What am I missing?

File: test-suite/success/search.v
Comment (Zimmi48) [inline test-suite/success/search.v:32]: Same question.

File: test-suite/stm/Nijmegen_QArithSternBrocot_Zaux.v
Comment (Zimmi48) [inline test-suite/stm/Nijmegen_QArithSternBrocot_Zaux.v:22]: `Omega` is already exported by `ZArith`.

**Comment by Zimmi48** [comment]
Also I would need at least an ack of [this comment](https://github.com/coq/coq/pull/11976#issuecomment-613505269) before merging.

**Comment by ejgallego** [comment]
> Also I would need at least an ack of [this comment](https://github.com/coq/coq/pull/11976#issuecomment-613505269) before merging.

In my view the benefits of starting the consolidation of arithmetic tactics are greater than the downsides, so even if `lia` is still not perfect (how could it be) I'd be in favor of starting the deprecation in 8.12; this is going to be a very long transition anyways.

File: doc/sphinx/proof-engine/vernacular-commands.rst
Comment (vbgl) [reply doc/sphinx/proof-engine/vernacular-commands.rst]: I’ve removed this change.

File: test-suite/success/search.v
Comment (vbgl) [reply test-suite/success/search.v]: same answer

File: test-suite/stm/Nijmegen_QArithSternBrocot_Zaux.v
Comment (vbgl) [reply test-suite/stm/Nijmegen_QArithSternBrocot_Zaux.v]: idem

File: doc/sphinx/addendum/omega.rst
Comment (Zimmi48) [inline doc/sphinx/addendum/omega.rst:17]: Sorry @vbgl for not seeing this earlier, but this paragraph should be reformulated as well. Now, we recommend switching all existing code to `lia` and reporting any issue. If no new issues are reported, `omega` will be removed soon.

**Review by Zimmi48 — COMMENTED** [review]
When this last comment is fixed, I intend to merge. So if someone has anything to add, they should do it now.

File: doc/sphinx/addendum/omega.rst
Comment (vbgl) [reply doc/sphinx/addendum/omega.rst]: Done.

**Review by Zimmi48 — APPROVED** [review]

**Comment by vbgl** [comment]
Thank you very much Théo!

**Comment by Zimmi48** [comment]
Thanks to you!

**Pull request merged**
